gp-res-filter/src/main/java/com/ibm/g11n/pipeline/resfilter/JsonResource.java

```diff
@@ -21,52 +21,70 @@
+    public Collection<ResourceString> parse(InputStream inStream) throws IOException {
+        Collection<ResourceString> resultCol = new LinkedList<ResourceString>();
+        try (InputStreamReader reader = new InputStreamReader(new BomInputStream(inStream), StandardCharsets.UTF_8)) {
             JsonElement root = new JsonParser().parse(reader);
             if (!root.isJsonObject()) {
                 throw new IllegalResourceFormatException("The root JSON element is not an JSON object.");
             }
+            int sequenceNum = 0;
             for (Map.Entry<String, JsonElement> entry : root.getAsJsonObject().entrySet()) {
                 String key = entry.getKey();
                 JsonElement value = entry.getValue();
                 if (!value.isJsonPrimitive() || !value.getAsJsonPrimitive().isString()) {
                     throw new IllegalResourceFormatException("The value of JSON element " + key + " is not a string.");
                 }
-                resultMap.put(key, value.getAsString());
+                ResourceString res = new ResourceString(key, value.getAsString());
+                sequenceNum++;
+                res.setSequenceNumber(sequenceNum);
+                resultCol.add(res);
             }
         } catch (JsonParseException e) {
             throw new IllegalResourceFormatException("Failed to parse the specified JSON contents.", e);
         }
-        return resultMap;
+        return resultCol;
     }
 
     @Override
-    public void write(OutputStream outStream, String language, Map<String, String> data) throws IOException {
-        try (OutputStreamWriter writer = new OutputStreamWriter(new BufferedOutputStream(outStream), "UTF-8")) {
-            new Gson().toJson(data, writer);
+    public void write(OutputStream outStream, String language, Collection<ResourceString> data) throws IOException {
+        // extracts key value pairs in original sequence order
+        TreeSet<ResourceString> sortedResources = new TreeSet<>(new ResourceStringComparator());
+        sortedResources.addAll(data);
+        LinkedHashMap<String, String> kvmap = new LinkedHashMap<>(sortedResources.size());
+        for (ResourceString res : sortedResources) {
+            kvmap.put(res.getKey(), res.getValue());
+        }
+        try (OutputStreamWriter writer = new OutputStreamWriter(new BufferedOutputStream(outStream),
+                StandardCharsets.UTF_8)) {
+            new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create().toJson(kvmap, writer);
         }
     }
 
     @Override
-    public void merge(InputStream base, OutputStream outStream, String language, Map<String, String> data) throws IOException{
+    public void merge(InputStream base, OutputStream outStream, String language, Collection<ResourceString> data)
+            throws IOException {
         throw new UnsupportedOperationException("Merging JSON resource is not supported.");
```

Comment:
could this, at least, be supported by `parse()/write()`? though it wouldn't be a completely clean merge.